File: limap/undistortion/undistort.py
import cv2
from _limap import _base, _undistortion


def undistort_image_camera(camera, imname_in, imname_out):
    """
    Run COLMAP undistortion on one single image with an input camera. \
    The undistortion is only applied if the camera model is \
    neither "SIMPLE_PINHOLE" nor "PINHOLE".

    Args:
        camera (:class:`limap.base.Camera`): \
            The camera (type + parameters) for the image.
        imname_in (str): filename for the input image
        imname_out (str): filename for the output undistorted image

    Returns:
        :class:`limap.base.Camera`: The undistorted camera
    """
    if camera.IsUndistorted():  # no distortion
        img = cv2.imread(imname_in)
        cv2.imwrite(imname_out, img)
        if int(camera.model) == 0 or int(camera.model) == 1:
            return camera
        # if "SIMPLE_RADIAL", update to "SIMPLE_PINHOLE"
        if int(camera.model) == 2:
            new_camera = _base.Camera(
                "SIMPLE_PINHOLE",
                camera.K(),
                cam_id=camera.camera_id,
                hw=[camera.h(), camera.w()],
            )
        else:
            # else change to pinhole
            new_camera = _base.Camera(
                "PINHOLE",
                camera.K(),
                cam_id=camera.camera_id,
                hw=[camera.h(), camera.w()],
            )
        return new_camera

    # undistort
    camera_undistorted = _undistortion._UndistortCamera(
        imname_in, camera, imname_out
    )
    return camera_undistorted


def undistort_points(points, distorted_camera, undistorted_camera):
    """
    Run COLMAP undistortion on the keypoints.

    Args:
        points (list[:class:`np.array`]): \
            List of 2D keypoints on the distorted image
        distorted_camera (:class:`limap.base.Camera`): \
            The camera before undistortion
        undistorted_camera (:class:`limap.base.Camera`): \
            The camera after undistortion

    Returns:
        list[:class:`np.array`]: \
            List of the corresponding 2D keypoints on the undistorted image
    """
    return _undistortion._UndistortPoints(
        points, distorted_camera, undistorted_camera
    )

File: limap/undistortion/undistort.cc
#include "undistortion/undistort.h"

#include <colmap/image/undistortion.h>
#include <colmap/scene/camera.h>
#include <colmap/sensor/bitmap.h>
#include <iostream>

namespace limap {

namespace undistortion {

Camera UndistortCamera(const std::string &imname_in, const Camera &camera,
                       const std::string &imname_out) {
  colmap::Bitmap img, img_undistorted;
  img.Read(imname_in);
  colmap::Camera cam = camera;

  bool exif_autorotate = false;
  if (cam.height != img.Height() || cam.width != img.Width()) {
    if (cam.width != img.Height() || cam.height != img.Width())
      throw std::runtime_error("Error! The height and width of the given "
                               "camera do not match the input image.");
    // std::cout<<"[WARNING] Auto rotating image (EXIF):
    // "<<imname_in<<std::endl;
    exif_autorotate = true;
    cam.Rescale(img.Width(), img.Height());
  }
  colmap::UndistortCameraOptions undist_options;
  colmap::Camera cam_undistorted;
  colmap::UndistortImage(undist_options, img, cam, &img_undistorted,
                         &cam_undistorted);
  img_undistorted.Write(imname_out);
  if (exif_autorotate)
    cam_undistorted.Rescale(img_undistorted.Height(), img_undistorted.Width());

  Camera camera_undistorted = cam_undistorted;
  return camera_undistorted;
}

CameraView UndistortCameraView(const std::string &imname_in,
                               const CameraView &view,
                               const std::string &imname_out) {
  CameraView new_view = view;
  new_view.cam = UndistortCamera(imname_in, view.cam, imname_out);
  return new_view;
}

V2D UndistortPoint(const V2D &point, const Camera &distorted_camera,
                   const Camera &undistorted_camera) {
  return undistorted_camera.ImgFromCam(distorted_camera.CamFromImg(point));
}

std::vector<V2D> UndistortPoints(const std::vector<V2D> &points,
                                 const Camera &distorted_camera,
                                 const Camera &undistorted_camera) {
  std::vector<V2D> new_points;
  for (auto it = points.begin(); it != points.end(); ++it) {
    new_points.push_back(
        UndistortPoint(*it, distorted_camera, undistorted_camera));
  }
  return new_points;
}

} // namespace undistortion

} // namespace limap
